### PR TITLE
First Implementation

### PR DESCRIPTION
**Pull request by rdohms**
This first implementation is feature complete.

- [x] reads current labels
- [x] defines which labels need to be created
- [x] defines which to be removed
- [x] defines which need to be renamed instead
- [x] defines which need to be updated

### CODE REVIEW

File: cli.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  .then(json => {
+    const manager = new LabelManager(
+      _.merge(options, {
+        labels: json.labels,
+        repository,
+        github,
+      })
+    );
+
+    manager.sync((err, operations) => {
```
[the remark this answers is not in the document]
Comment (rdohms) [reply cli.js]: @spirosikmd i love the idea, how would `sync` look like in a way where it returns operations? I'm still a bit confused on how to do that in combination with Promise.all and/or using async.

[the remark this answers is not in the document]
Comment (spirosikmd) [reply cli.js]: It is almost the same idea. I changed the structure of the return value to be able to either return operations or error details and operations.

```js
/*
 * We declare this function as async (async sync makes it funny. I suggest we keep it like this :)
 * When we declare a function as `async` it returns a promise. We can still use the function as
 *
 * sync()
 *   .then(({ operations }) => { ... })
 *   .catch({ details, operations } => { ... })
 */
async sync() {
  const existingLabels = await this.githubLabels.getLabels();

  const currentLabels = new LabelCollection(existingLabels || []);
  const futureLabels = new LabelCollection(this.newLabels);
  const operations = currentLabels.compareCollections(futureLabels);

  if (this.dryRun) {
    return { operations };
  }

  try {
    await Promise.all([
      this.githubLabels.update(operations.replace),
      this.githubLabels.create(operations.create),
      this.githubLabels.delete(operations.delete),
    ]);

    return { operations };
  } catch (error) {
    const details = JSON.parse(err.toString());
    return { details, operations };
  }
}
```

To me, this is easier to understand!

[the remark this answers is not in the document]
Comment (rdohms) [reply cli.js]: you are absolutely right, this was what i needed to understand async/await and i love it. Working on rewriting stuff based on your great ideas.

**Comment by rdohms** [comment]
@spirosikmd regarding testing the manager, yep need to do it, but the link you showed about mocking only has examples with the use of `import ... from` and I use `require` everywhere cause i could not get the other stuff running nicely, and with that i really can't seem to get the mocking right.

**Comment by spirosikmd** [comment]
@rdohms I tried and this works:

```js
const LabelManager = require('../src/label-manager');

jest.mock('../src/client/labels-api');

describe('LabelManager', () => {
  let labelManager;

  beforeEach(() => {
    labelManager = new LabelManager({
      dryRun: false,
    });
  });

  it('passes', () => {
    console.log(labelManager.githubLabels);
  });
});
```

Jest mocked `labelManager.githubLabels` so know you can mock its returned values.

**Comment by rdohms** [comment]
@spirosikmd you rock dude, thank you so much for the light, i made the changes you recommended, might even make some more now that i have the hang of it.

**Comment by rdohms** [comment]
@spirosikmd however the lint is rejecting the async/await, which rule am I missing is this later then es2016?

**Comment by rdohms** [comment]
doh, yeah nevermind, fixed it

File: cli.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  config: cli.flags.config,
+};
+
+if (cli.flags.dryrun) {
+  console.log(chalk.yellow('Dry run, no labels are copied for real:'));
+}
+
+loadJsonFile(path.normalize(options.config))
+  .then(json => {
+    const [owner, repo] = repositoryParser.parse(repository);
```
Comment (spirosikmd) [inline cli.js]: Should we run this directly after line 60, so we break soon if the provided repository argument is not correct?

File: cli.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  dryRun: cli.flags.dryrun,
+  token: cli.flags.token,
+  config: cli.flags.config,
+};
+
+if (cli.flags.dryrun) {
+  console.log(chalk.yellow('Dry run, no labels are copied for real:'));
+}
+
+loadJsonFile(path.normalize(options.config))
```
Comment (spirosikmd) [inline cli.js]: You can still use `async/await` here. For example

```js
async function run() {
  try {
    const json = await loadJsonFile(path.normalize(options.config));

    const syncer = new LabelSyncer({
      ...options,
      ...{ owner, repo, github },
    });

    try {
      const operations = await syncer.sync(labels.build(json.labels));

      // log the operation messages...
    } catch (error) {
      // handle error
    }
  } catch (error) {
    // handle error
  }
}

run();
```

File: package.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    "jest": "^23.5.0",
+    "prettier": "^1.14.2"
+  },
+  "dependencies": {
+    "@octokit/rest": "^16",
+    "chalk": "^2.4.1",
+    "load-json-file": "^5.0.0",
+    "lodash": "^4.17.10",
+    "meow": "^5.0.0",
+    "proxyquire": "^2.1.0"
```
Comment (spirosikmd) [inline package.json]: Are we using `proxyrequire`? If not, we can remove it!

File: package.json [hunk 5]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "@rdohms/gh-labeller",
+  "version": "0.1.0",
+  "description": "CLI Command that creates GitHub Labels from a set of configuration",
+  "main": "cli.js",
+  "license": "MIT",
+  "devDependencies": {
+    "eslint": "^4.19.1",
+    "eslint-config-prettier": "^3.3.0",
+    "eslint-plugin-import": "^2.12.0",
```
Comment (spirosikmd) [inline package.json]: Are we using this plugin? If not, we can remove it!

File: src/client/labels-api.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+          owner: this.owner,
+          repo: this.repo,
+          name: label.name,
+        });
+      })
+    );
+  }
+
+  /**
+   * @param {String} fullname
```
Comment (spirosikmd) [inline src/client/labels-api.js]: What is this comment here?

File: src/client/labels-api.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+    return labels.map(label => new Label(label));
+  }
+
+  /**
+   * @param {Object} operations
+   * @returns {Promise}
+   */
+  update(operations) {
+    if (this.dryRun) {
```
Comment (spirosikmd) [inline src/client/labels-api.js]: Why are we using the dry run here? In the [syncer](https://github.com/rdohms/gh-labeller/pull/1/files#diff-e02504d72320c34de73a781c47a29e89R18) if we specified dry run, we return immediately, and we don't execute these functions.

File: src/client/labels-api.js [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+   */
+  async getLabels() {
+    const labels = await this.github.paginate(
+      this.github.issues.listLabelsForRepo.endpoint.merge({
+        owner: this.owner,
+        repo: this.repo,
+      })
+    );
+
+    return labels.map(label => new Label(label));
```
Comment (spirosikmd) [inline src/client/labels-api.js]: We could call the `buildLabels` function here as well.

File: src/model/labels.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    matching.forEach(matchingLabel => {
+      if (matchingLabel.hash() !== newLabel.hash()) {
+        toUpdate.push({
+          oldLabel: matchingLabel,
+          newLabel,
+        });
+      }
+    });
+
+    if (matching.length !== 0) {
```
Comment (spirosikmd) [inline src/model/labels.js]: Should we check this directly after line 15 so we can exit early?

File: package.json (hunk 4 above)
Comment (rdohms) [reply package.json]: aha! i was ahead of you there :P

File: package.json (hunk 5 above)
Comment (rdohms) [reply package.json]: not sure to be honest, copied this one over from our old code

File: src/model/labels.js (hunk 9 above)
Comment (rdohms) [reply src/model/labels.js]: nope, cause the method above this adds things to the toUpdate array, so we need to execute that one.

File: package.json (hunk 5 above)
Comment (rdohms) [reply package.json]: yeah confirmed we do not need it

File: src/model/labels.js (hunk 9 above)
Comment (spirosikmd) [reply src/model/labels.js]: Oh yeah, you are right! Sorry I misread that as `matching.length === 0`

**Comment by rdohms** [comment]
Alright, i think this is it @spirosikmd! Unless i have fucked up something else?

**Comment by rdohms** [comment]
@spirosikmd thank you very much for the guidance, could not have done it without your help. Only makes me miss working with you more 😄

**Comment by spirosikmd** [comment]
I enjoyed that a lot! I am glad I was able to help 😄

**Pull request merged**